File: src/algorithms/python/Combination_Sum_III.py
# -*- coding: utf-8 -*-


class Solution:
    def dfs(self, num, left, cnt):
        if cnt == 0:
            if left == 0:
                self.ret.append(list(self.now))
            return
        for i in range(num + 1, 10):
            self.now.append(i)
            self.dfs(i, left - i, cnt - 1)
            self.now.pop()

    # @param {integer} k
    # @param {integer} n
    # @return {integer[][]}
    def combinationSum3(self, k, n):
        self.ret, self.now = [], []
        self.dfs(0, n, k)
        return self.ret


if __name__ == "__main__":
    solution = Solution()
    print solution.combinationSum3(3, 7)
    print solution.combinationSum3(3, 9)

File: src/algorithms/cpp/Combination_Sum_III.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
private:
    vector <vector<int>> result;
    vector <int> tmp;
    int n;

private:
    void dfs(int now, int sum, int left) {
        if (left <= 0) {
            if (sum == n)   result.push_back(tmp);
            return;
        }
        if (now == 9)       return;
        for (int i = now + 1; i <= 9; ++ i) {
            tmp.push_back(i);
            dfs(i, i + sum, left - 1);
            tmp.pop_back();
        }
    }

public:
    vector<vector<int>> combinationSum3(int k, int n) {
        this->n = n;
        for (int i = 1; i <= 9 - k + 1; ++ i) {
            tmp.push_back(i);
            dfs(i, i, k - 1);
            tmp.pop_back();
        }
        return result;
    }
};

int main() {
    Solution solution;
    vector <vector<int>> result = solution.combinationSum3(9, 45);
    for (auto v: result) {
        for (auto num: v)   cout << num << " ";
        cout << endl;
    }
    return 0;
}
